src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -502,13 +501,18 @@ static void handle_parallel_super_load(Symbol* name,
                                        Handle class_loader,
                                        Handle protection_domain, TRAPS) {
 
-  // superk is not used; resolve_super_or_fail is called for circularity check only.
-  Klass* superk = SystemDictionary::resolve_super_or_fail(name,
-                                                          superclassname,
-                                                          class_loader,
-                                                          protection_domain,
-                                                          true,
-                                                          CHECK);
+  // The result superk is not used; resolve_with_circularity_detection is called for circularity check only.
+  // This passes true to is_superclass even though it might not be the super class in order to perform the
+  // optimization anyway.
+  // The only thing that takes different action for is_superclass is dumping the static archive, which doesn't
+  // reach this path.
+  assert (!CDSConfig::is_dumping_static_archive(), "should not be dumping static archive");
```

Comment:
I don't think this assert is correct.

It may be true that *some* calls made by CDS will not come to here (i.e., the code with the `// Special processing for handling UNREGISTERED shared classes` comment).

However, it's not guaranteed that EVERY call made when dumping the static archive will not come to here.

Reply:
This is why I wanted you to look at it.  If you follow this code, it's calling 

      // Special processing for handling UNREGISTERED shared classes.
      InstanceKlass* k = SystemDictionaryShared::lookup_super_for_unregistered_class(class_name,
                             next_name, is_superclass);

With is_superclass is true, which does this.  which is wrong:

    if (class_name->equals(parser->current_class_name())) {
      // When this function is called, all the numbered super and interface types
      // must have already been loaded. Hence this function is never recursively called.
      if (is_superclass) {
        return parser->lookup_super_for_current_class(super_name);
      } else {
        return parser->lookup_interface_for_current_class(super_name);
      }

Maybe it's not the parsing_thread, whatever that means and doesn't get here.  I can remove the assert, which is the pre-existing code, but file an bug to look at this.